### PR TITLE
Invocations filter

### PR DESCRIPTION
**Pull request by IDoneShaveIt**
1. Support invocation filter for the report
2. Refactor data monitoring 
3. Removed the dependencies between the report test results and the alerts for tests

Needs to do:
1. Docs
2. Add tests
3. Update the bundle

### CODE REVIEW

File: elementary/monitor/cli.py [hunk 1]
```diff
@@ -102,7 +107,7 @@ def get_cli_properties() -> dict:
 @common_options
 @click.option(
     "--slack-webhook",
-    "-s",
+    "-sw",
```
Comment (IDoneShaveIt) [inline elementary/monitor/cli.py:110]: We want to use -s as dbt does for selecting node

File: elementary/monitor/data_monitoring/data_monitoring.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+from elementary.utils import package
+from elementary.utils.log import get_logger
+
+logger = get_logger(__name__)
+
+YAML_FILE_EXTENSION = ".yml"
+SQL_FILE_EXTENSION = ".sql"
+
+
+class DataMonitoring:
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring.py:22]: Split DataMonitoring into 3:
1. DataMonitoring - base class for data monitoring
2. DataMonitoringAlerts - Responsible for our alerts
3. DataMonitoringReport - Responsible for our report

We got to a point where we have a too much of unnecessary coupling between the report and the alerts.
Those are 2 different features and should not be coupled.

Comment (IDoneShaveIt) [reply elementary/monitor/data_monitoring/data_monitoring.py:22]: No new method, only moved the existing ones to a new file!

File: elementary/monitor/data_monitoring/data_monitoring_alerts.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+from elementary.utils.json_utils import prettify_json_str_set
+from elementary.utils.log import get_logger
+
+logger = get_logger(__name__)
+
+YAML_FILE_EXTENSION = ".yml"
+SQL_FILE_EXTENSION = ".sql"
+
+
+class DataMonitoringAlerts(DataMonitoring):
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_alerts.py:28]: No new methods, just moved to a different file

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+from elementary.utils.log import get_logger
+from elementary.utils.time import get_now_utc_iso_format
+
+logger = get_logger(__name__)
+
+YAML_FILE_EXTENSION = ".yml"
+SQL_FILE_EXTENSION = ".sql"
+
+
+class DataMonitoringReport(DataMonitoring):
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:39]: Did a bit of an order in the mess we got here.
Added support for invocation filter as well

File: elementary/clients/api/api.py [hunk 5]
```diff
@@ -1,6 +1,16 @@
+from collections import defaultdict
+from typing import Any
+
 from elementary.clients.dbt.dbt_runner import DbtRunner
 
 
 class APIClient:
     def __init__(self, dbt_runner: DbtRunner):
         self.dbt_runner = dbt_runner
+        self.run_cache = defaultdict(lambda: None)
```
Comment (IDoneShaveIt) [inline elementary/clients/api/api.py:10]: Added a run cache - now we want have to pass all the API results for each API call that is relying on it, and we can make a mistakable second operation run to not be executed :)

File: elementary/monitor/api/invocations/invocations.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+from elementary.clients.api.api import APIClient
+from elementary.monitor.api.invocations.schema import DbtInvocationSchema
+from elementary.utils.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class InvocationsAPI(APIClient):
+    def get_last_invocation(self, type: str) -> DbtInvocationSchema:
```
Comment (IDoneShaveIt) [inline elementary/monitor/api/invocations/invocations.py:11]: It felt right to me to split those methods into 3 different methods.
If you feel like this is over engineering (or code duplicating), I can merge those 3

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+                local_html_path, remote_bucket_file_path=remote_file_path
+            )
+            self.execution_properties["sent_to_gcs_successfully"] = send_succeded
+            if not send_succeded:
+                self.success = False
+
+        self.execution_properties["success"] = self.success
+        return self.success
+
+    def _get_lineage(self, exclude_elementary_models: bool = False) -> LineageSchema:
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:235]: old code

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+            return dict(), dict()
+
+    @staticmethod
+    def _serialize_totals(totals: Dict[str, TotalsSchema]) -> Dict[str, dict]:
+        serialized_totals = dict()
+        for model_unique_id, total in totals.items():
+            serialized_totals[model_unique_id] = total.dict()
+        return serialized_totals
+
+    def _get_models_runs_and_totals(self, models_runs: List[ModelRunsSchema]):
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:294]: old code

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+            models_runs_dicts.append(model_runs.dict(by_alias=True))
+            model_runs_totals[model_runs.unique_id] = {
+                "errors": model_runs.totals.errors,
+                "warnings": 0,
+                "failures": 0,
+                "passed": model_runs.totals.success,
+            }
+        return models_runs_dicts, model_runs_totals
+
+    def _get_dbt_models_and_sidebars(
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:307]: old code

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+            serializable_nodes[key] = dict(nodes[key])
+
+        # Currently we don't show exposures as part of the sidebar
+        sidebars = self.sidebar_api.get_sidebars(
+            artifacts=[*models.values(), *sources.values()]
+        )
+
+        return serializable_nodes, sidebars
+
+    def _get_dbt_models_test_coverages(self) -> Dict[str, Dict[str, int]]:
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:329]: old code

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        coverages = self.models_api.get_test_coverages()
+        return {model_id: dict(coverage) for model_id, coverage in coverages.items()}
+
+    def _get_report_file_path(self, file_path: Optional[str] = None) -> str:
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:333]: old code

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+            try:
+                webbrowser.open_new_tab("file://" + html_path)
+            except webbrowser.Error:
+                logger.error("Unable to open the web browser.")
+
+        self.execution_properties["report_end"] = True
+        self.execution_properties["success"] = self.success
+        return self.success, html_path
+
+    def send_report(
```
Comment (IDoneShaveIt) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:205]: old code

File: elementary/monitor/api/tests/tests.py [hunk 13]
```diff
@@ -28,34 +52,207 @@ def get_test_sub_type_unique_id(
...
+        if filter.invocation_id:
+            invocation = self.invocatons_api.get_invocation_by_id(
+                type="test", invocation_id=filter.invocation_id
+            )
+        elif filter.invocation_time:
+            invocation = self.invocatons_api.get_invocation_by_time(
+                type="test", invocation_max_time=filter.invocation_time
+            )
+        elif filter.last_invocation:
+            invocation = self.invocatons_api.get_last_invocation(type="test")
```
Comment (haritamar) [inline elementary/monitor/api/tests/tests.py:87]: If none of these filters exists, it means there's no filter by invocation?
(if that's the case, worth adding a comment that clarifies it)

File: elementary/monitor/cli.py [hunk 14]
```diff
@@ -147,11 +152,18 @@ def get_cli_properties() -> dict:
...
     type=str,
     default=None,
     help="A slack webhook URL for sending alerts to a specific channel.",
 )
+@click.option(
+    "--deprecated-slack-webhook",
+    "-s",  # Deprecatred - will be used for --select in the future
```
Comment (haritamar) [inline elementary/monitor/cli.py:162]:
```suggestion
    "-s",  # Deprecated - will be used for --select in the future
```

File: elementary/monitor/data_monitoring/data_monitoring_report.py [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+        self.s3_client = S3Client.create_client(self.config, tracking=self.tracking)
+        self.gcs_client = GCSClient.create_client(self.config, tracking=self.tracking)
+
+    def _parse_filter(self, filter: Optional[str] = None) -> DataMonitoringFilter:
+        data_monitoring_filter = DataMonitoringFilter()
+        if filter:
+            invocation_id_regex = re.compile(r"invocation_id:\w+")
+            invocation_time_regex = re.compile(
+                r"invocation_time:([\+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24\:?00))?(\17[0-5]\d)))"
+            )
```
Comment (haritamar) [inline elementary/monitor/data_monitoring/data_monitoring_report.py:65]: Do we actually need the full huge regex?
Can we maybe only rely on the "invocation_id:" or "invocation_time:" prefix instead?

I think it makes sense to first parse the prefix, and then if needed show an error that's specific to the prefix used (e.g. "Invalid timestamp provided for invocation_time selector", which you can get from regular datetime parsing)

File: elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql [hunk 16]
```diff
@@ -1,21 +1,23 @@
...
 {% endmacro %}
 
 
-{% macro current_tests_run_results(days_back = none) %}
+{% macro current_tests_run_results(days_back = none, invocation_id = none) %}
     with elementary_test_results as (
         select * from {{ ref('elementary', 'elementary_test_results') }}
-        {% if days_back %}
+        {% if invocation_id %}
+            where invocation_id = {{ "'" ~ invocation_id ~ "'"}}
```
Comment (haritamar) [inline elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql:19]: Not critical, but Elon convinced me that using python string formatting is clearer than the Jinja concat syntax :)
(e.g. `'{}'.format(invocation_id)`)

File: elementary/monitor/data_monitoring/schema.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+    invocation_id: Optional[str] = None
+    invocation_time: Optional[str] = None
+    last_invocation: Optional[bool] = False
+
+    @validator("invocation_time", pre=True)
+    def format_invocation_time(cls, invocation_time):
+        if invocation_time:
+            invocation_datetime = convert_local_time_to_timezone(
+                datetime.fromisoformat(invocation_time)
+            )
```
Comment (haritamar) [inline elementary/monitor/data_monitoring/schema.py:19]: When documenting this feature - it's worth noting that the invocation_time is expected in local time

File: elementary/monitor/api/invocations/invocations.py (hunk 6 above)
Comment (IDoneShaveIt) [reply elementary/monitor/api/invocations/invocations.py:11]: I like it 🙂

File: elementary/monitor/api/tests/tests.py [hunk 18]
```diff
@@ -28,34 +52,207 @@ def get_test_sub_type_unique_id(
...
-        )
-        return tests_metrics
+                disable_samples=disable_samples,
+            )
+
+        test_results = defaultdict(list)
+        for test_metadata in test_results_metadata:
+            test_metadata_dict = dict(test_metadata)
+            test_sub_type_unique_id = self.get_test_sub_type_unique_id(
+                **dict(test_metadata_dict)
```
[the remark this answers is not in the document]
Comment (IDoneShaveIt) [reply elementary/monitor/api/tests/tests.py:216]: You are right!

File: elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql (hunk 16 above)
Comment (IDoneShaveIt) [reply elementary/monitor/dbt_project/macros/base_queries/current_tests_run_results_query.sql:19]: Agree!

File: elementary/monitor/data_monitoring/schema.py (hunk 17 above)
Comment (IDoneShaveIt) [reply elementary/monitor/data_monitoring/schema.py:19]: Will add! (the documentation is already merged - was part of the old docs branch 🙂 )

**Review by haritamar — APPROVED** [review]

**Pull request merged**